sql/ttl: refactor progress tracking behind interface for future checkpointing

spilchen opened the pull request:
This change introduces a progressTracker interface to abstract how TTL job progress is tracked. The existing behavior is now encapsulated in a legacyProgressTracker implementation.

While TTL jobs could previously resume after interruption, they would always start from the beginning of the table, effectively discarding any partial progress. This refactor prepares for future support of span-level progress tracking, enabling jobs to resume from where they left off.

A follow-up change will add actual checkpointing support using this interface, allowing resumable TTL jobs to skip spans already processed.

This commit is scoped to the interface definition and refactoring of the current implementation under it.

Informs #140514

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151243)
<!-- Reviewable:end -->

spilchen committed:
sql/ttl: refactor progress tracking behind interface for future checkpointing

This change introduces a progressTracker interface to abstract how TTL job
progress is tracked. The existing behavior is now encapsulated in a
legacyProgressTracker implementation.

While TTL jobs could previously resume after interruption, they would always
start from the beginning of the table, effectively discarding any partial
progress. This refactor prepares for future support of span-level progress
tracking, enabling jobs to resume from where they left off.

A follow-up change will add actual checkpointing support using this interface,
allowing resumable TTL jobs to skip spans already processed.

This commit is scoped to the interface definition and refactoring of the current
implementation under it.

Informs #140514

Release note: none

rafiss reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151243#-:-OX0XN-S0I2gR5V07kxR:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
